### PR TITLE
to be able show connection failed message

### LINKED ISSUE #404: Suggestion for displaying network errors In ContributorsScreen and StaffScreen
## Overview
I thought it would be nice to be able to give out something other than the title when the network has an error, such as offline...!
(Nothing was shown for 2023 as well as the present when offline)

|ContributorsScreen|StaffScreen|
|---|---|
|||

## Ref
- With [NowInAndroid](https://github.com/android/nowinandroid) , the Snackbar, which never disappears, was always displayed to show that you were offline.

### PR DESCRIPTION
**Pull request by kosenda**
## Issue
- close #404
  - I have not been assigned to an issue, but I thought that no one but me would show up to do it, so I created this PR.

## Overview (Required)
- Created `knownPlatformExceptionOrNull` and `InternetConnectionException` to match The following [comments](https://github.com/DroidKaigi/conference-app-2024/issues/404#issuecomment-2290206708) 

>How about adding a expect function like this?
>
>expect fun knownPlatformExceptionOrNulll(e: Exception): AppError?
>
>actual fun knownPlatformExceptionOrNulll(e: Exception): AppError? {
>if(e is UnknownHostException) {
>return AppError.InternetConnectionException(e)
>}
>return null
>}

- Composable function `ComposeResourceErrorMessages` to display network connection failure in English or Japanese
- Error wording in Japanese and English is my own wording

## Movie
Before | After
:--: | :--:
<video src="https://github.com/user-attachments/assets/9d10f9ff-863c-439d-8d9e-f2ee851e24bd" width="300" > | <video src="https://github.com/user-attachments/assets/92abe38f-c1f9-448b-be8a-3b7e2c2326c4" width="300" >

### CODE REVIEW

File: core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/ComposeResourceErrorMessage.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import androidx.compose.runtime.Composable
+import conference_app_2024.core.ui.generated.resources.connection_failed
+import org.jetbrains.compose.resources.stringResource
+
+/**
+ * If you want to display composeResource error messages instead of error messages from the API, define them here
+ * [Throwable.toApplicationErrorMessage] also needs to be defined
+ */
+@Composable
+internal fun ComposeResourceErrorMessages(): List<ComposeResourceErrorMessage> = listOf(
```
Comment (takahirom) [inline core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/ComposeResourceErrorMessage.kt:12]: Could you use `*c*omposeResourceErrorMessages` for the name?
https://github.com/androidx/androidx/blob/androidx-main/compose/docs/compose-api-guidelines.md#naming-composable-functions-that-return-values

File: core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/ComposeResourceErrorMessage.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    ComposeResourceErrorMessage(
+        ComposeResourceErrorMessageType.ConnectionFailed,
```
Comment (takahirom) [inline core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/ComposeResourceErrorMessage.kt:14]: Can we have a AppError.InternetConnectionException::class here?

File: core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/ComposeResourceErrorMessage.kt (hunk 1 above)
Comment (kosenda) [reply core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/ComposeResourceErrorMessage.kt]: Sorry for the lack of reading the guidelines as I did yesterday 🙇

File: core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/ComposeResourceErrorMessage.kt (hunk 2 above)
Comment (kosenda) [reply core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/ComposeResourceErrorMessage.kt]: KClass I had never used it myself and couldn't picture it...
I fixed it by using `AppError.InternetConnectionException::class`!

File: core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/ComposeResourceErrorMessage.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+@Composable
+internal fun composeResourceErrorMessages(): List<ComposeResourceErrorMessage> = listOf(
+    ComposeResourceErrorMessage(
+        InternetConnectionException::class,
+        stringResource(UiRes.string.connection_failed),
+    ),
+)
+
+internal data class ComposeResourceErrorMessage(
+    val appError: KClass<out AppError>,
```
Comment (takahirom) [inline core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/ComposeResourceErrorMessage.kt:22]: Thanks. Can we rename this to `appErrorClass`? 🙏

File: core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/AppErrorMapper.kt [hunk 4]
```diff
@@ -1,5 +1,12 @@
 package io.github.droidkaigi.confsched.ui
 
-fun Throwable.toApplicationErrorMessage(): String {
+import io.ktor.util.reflect.instanceOf
+
+internal fun Throwable.toApplicationErrorMessage(
+    composeResourceErrorMessages: List<ComposeResourceErrorMessage>? = null,
+): String {
+    composeResourceErrorMessages?.forEach {
+        if (this.instanceOf(it.appError)) return it.message
```
Comment (takahirom) [inline core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/AppErrorMapper.kt:9]: Thanks. Can we use `it. appErrorClass.isInstance(this)`?

Comment (kosenda) [reply core/ui/src/commonMain/kotlin/io/github/droidkaigi/confsched/ui/AppErrorMapper.kt]: I did not know `isInstance`! This one has been fixed as well!

**Review by takahirom — APPROVED** [review]
Looks great! You nailed it!

**Pull request merged**